qa/zcash/updatecheck.py

```diff
@@ -0,0 +1,347 @@
+                { "v4.3.1": (4, 3, 1), "v4.2.0-rc1": None }),
+            DependsVersionGetter("zeromq")),
+        Dependency("leveldb",
+            GithubTagReleaseLister("google", "leveldb", "^v(\d+)\.(\d+)$",
+                { "v1.13": (1, 13) }),
+            LevelDbVersionGetter()),
+    ]
+
+    # Rust crates.
+    crates = [
+        "aes", "aesni", "aes_soft", "arrayvec", "bellman",
+        "arrayref", "autocfg", "bech32", "blake2b_simd", "blake2s_simd",
+        "bit_vec", "block_cipher_trait", "byteorder",
+        "block_buffer", "block_padding", "c2_chacha", "cfg_if",
+        "byte_tools", "constant_time_eq", "crossbeam", "digest", "fpe",
+        "crossbeam_channel", "crossbeam_deque", "crossbeam_epoch",
+        "crossbeam_utils", "crossbeam_queue", "crypto_api", "crypto_api_chachapoly",
+        "directories", "fake_simd", "getrandom", "hex", "log",
+        "futures_cpupool", "futures",
+        "generic_array", "lazy_static", "libc", "nodrop", "num_bigint",
+        "memoffset", "ppv_lite86", "proc_macro2", "quote",
+        "num_cpus", "num_integer", "num_traits", "opaque_debug", "pairing",
+        "rand", "typenum",
+        "rand_chacha", "rand_core", "rand_hc", "rand_os", "rand_xorshift",
+        "rustc_version", "scopeguard", "semver", "semver_parser", "sha2", "syn",
+        "unicode_xid", "wasi",
+        "winapi_i686_pc_windows_gnu", "winapi", "winapi_x86_64_pc_windows_gnu",
+    ]
+
+    for crate in crates:
+        dependencies.append(
+            Dependency("crate_" + crate,
+                RustCrateReleaseLister(crate),
+                DependsVersionGetter("crate_" + crate)
+            )
+        )
+
+    return dependencies
+
+class Version(list):
+    def __init__(self, version_tuple):
+        for part in version_tuple:
+            if part: # skip None's which can come from optional regexp groups
+                if str(part).isdigit():
+                    self.append(int(part))
+                else:
+                    self.append(part)
+
+    def __str__(self):
+        return '.'.join(map(str, self))
+
+    def __hash__(self):
+        return hash(tuple(self))
+
+class Dependency:
+    def __init__(self, name, release_lister, current_getter):
+        self.name = name
+        self.release_lister = release_lister
+        self.current_getter = current_getter
+
+    def current_version(self):
+        return self.current_getter.current_version()
+
+    def known_releases(self):
+        return sorted(self.release_lister.known_releases())
+
+    def released_versions_after_current_version(self):
+        current_version = self.current_version()
+        releases_after_current = []
+
+        for release in self.known_releases():
+            if release > current_version:
+                releases_after_current.append(release)
+
+        return releases_after_current
+
+    def is_up_to_date(self):
+        return len(self.released_versions_after_current_version()) == 0
```

Comment:
There's no caching here, so in `__main__` when we call `is_up_to_date()` followed by `known_releases()`, these trigger two separate network queries, which means:
- We hit external URLs more frequently (so CI might trigger rate limiting more easily).
- The script takes longer to complete.
- The two network queries might return completely different data, making the printed output for outdated dependencies unreliable (though the fact they are outdated could still be relied on).

I think it would make sense to cache the result of `self.release_lister.known_releases()` inside `Dependency`. Since this script is not long-running, the cache is "invalidated" by the script ending.